python/src/equistore/operations/join.py

```diff
@@ -0,0 +1,212 @@
+from typing import List
+
+import numpy as np
+
+from ..block import TensorBlock
+from ..labels import Labels
+from ..tensor import TensorMap
+from . import _dispatch
+from ._utils import _check_same_keys
+
+
+def join(tensormaps: List[TensorMap], axis: str):
+    """Join a sequence of :py:class:`TensorMap` along an axis.
+
+    The ``axis`` parameter specifies the type join.
+    For example, if ``axis='properties'`` it will be the joined along the
+    properties of the TensorMaps and if ``axis='samples'`` it will be the
+    along the samples.
+
+    :param tensormaps: sequence of :py:class:`TensorMap` for join
+    :param axis: A string indicating how the tensormaps are stacked.
+                 Allowed values are ``'properties'`` or ``'samples'``.
+
+    :return: The stacked :py:class:`TensorMap` which has more properties or
+             samples than the input TensorMap.
+    """
+    if axis not in ["properties", "samples"]:
+        raise ValueError("tensormaps can only joined along 'properties' or samples.")
+
+    stacked = tensormaps[0]
+
+    for to_stack in tensormaps[1:]:
+        _check_same_keys(stacked, to_stack, "join")
+
+        blocks = []
+        for key, block1 in stacked:
+            kwargs = dict(block1=block1, block2=to_stack.block(key))
+            if axis == "properties":
+                blocks.append(_join_blocks_along_properties(**kwargs))
+            elif axis == "samples":
+                blocks.append(_join_blocks_along_samples(**kwargs))
+            else:
+                raise ValueError(
+                    "Only `'properties'` or `'samples'` are "
+                    "valid values for the `axis` parameter."
+                )
+
+        stacked = TensorMap(stacked.keys, blocks)
+
+    return stacked
+
+
+def _join_labels(Labels1: Labels, Labels2: Labels):
+    """Join two of :py:class:`Labels`
+
+    Possible name and value clasehd are resolve by the following strategy:
+
+    1. If the names are the same and the values are unqiue we simply join them.
+       In example::
+
+        properties_1 = ["n"], [0, 2, 3]   |=> properties = ["n"], [0, 2, 3, 1, 4, 5]
+        properties_2 = ["n"], [1, 4, 5]   |
+
+    2. If the names are the same but the values are not unique a new property 
```

Comment:
most of the code uses the name `variable` for this

```suggestion
    2. If the names are the same but the values are not unique a new variable
```